### PR TITLE
Fix distribution policy for inherited tables

### PR DESCRIPTION
**Pull request by darthunix**
Current parser algorithm for setting distribution key column numbers
in gp_distribution_policy relation for inherited tables in GP5X
is doing something wrong at the moment (for GP6X everything is ok).
Query planes with GPORCA caused segmentation fault because of
out of range column numbers, Postgres optimizer simply
returned error before current patch. For example:
```
create table ta (a int) distributed randomly;
create table tb (a int, b int) inherits (ta) distributed by (b);

set optimizer=on;
insert into tb values(0, 0);
-- Segmentation fault

set optimizer=off;
insert into tb values(0, 0);
-- ERROR: no tlist entry for key 3 (cdbmutate.c:1484)

select attrnums from gp_distribution_policy where localoid = 'tb'::regclass;
 attrnums
----------
{3}
(1 row)
```
Also a check for setting non-hashable distribution key from a
parent table in an inherited one didn't work.
```
create table tc (a point) distributed randomly;
create table td (b int) inherits (tc) distributed by (a);

select * from td;
ERROR:  could not find mergejoinable = operator for type 600 (pathkeys.c:1174)
```
Co-authored-by: Vasiliy Ivanov <ivi@arenadata.io>

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@darthunix Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8364)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8364) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@darthunix Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8364)!

File: src/backend/parser/parse_utilcmd.c [hunk 1]
```diff
@@ -1694,100 +1693,157 @@ transformDistributedBy(ParseState *pstate, CreateStmtContext *cxt,
...
-				* In the ALTER TABLE case, don't complain about index keys
-				* not created in the command; they may well exist already.
-				* DefineIndex will complain about them if not, and will also
-				* take care of marking them NOT NULL.
-				*/
+				 * In the ALTER TABLE case, don't complain about index keys
+				 * not created in the command; they may well exist already.
+				 * DefineIndex will complain about them if not, and will also
+				 * take care of marking them NOT NULL.
+				 */
```
Comment (gkokolatos) [inline src/backend/parser/parse_utilcmd.c]: Unnecessary white space change.

Apologies for being pedantic about it. Simply when going through history to find something, it is confusing and makes things a bit more difficult to have a mix of inconsequential changes with actual code that one can not skip.

File: src/backend/parser/parse_utilcmd.c [hunk 2]
```diff
@@ -1802,7 +1858,6 @@ transformDistributedBy(ParseState *pstate, CreateStmtContext *cxt,
 
 	*policyp = policy;
 
-
```
Comment (gkokolatos) [inline src/backend/parser/parse_utilcmd.c]: same

File: src/backend/parser/parse_utilcmd.c [hunk 3]
```diff
@@ -1694,100 +1693,157 @@ transformDistributedBy(ParseState *pstate, CreateStmtContext *cxt,
...
+								   errmsg("type \"%s\" can't be a part of a "
+										  "distribution key",
+										  format_type_be(col->typeOid))));
 						}
+						elog(DEBUG1, "Distribution key \"%s\" matched a column \"%s\" with number %d in a column list",
+							 key, col->columnName, colindex);
+						found = true;
+
+						break;
 					}
```
Comment (gkokolatos) [inline src/backend/parser/parse_utilcmd.c]: I think the logic is fine.

However, the sam(ish) action is performed in three different blocks which reduces readability. How about introducing a small static helper to find key in list and another to add key in list?

File: src/backend/commands/tablecmds.c [hunk 4]
```diff
@@ -1599,26 +1599,7 @@ MergeAttributes(List *schema, List *supers, bool istemp, bool isPartitioned,
...
-					{
-						/* Check if any distribution attribute has higher offset than the current */
-						if (policy->attrs[attr_ofst] > child_attno)
-						{
-							Assert(policy->attrs[attr_ofst] > 0 && "index should not become negative");
-							policy->attrs[attr_ofst]--;
-						}
-					}
-				}
+				Assert(policy->nattrs >= 0 && "the number of distribution attributes is not negative");
```
[the remark this answers is not in the document]
Comment (darthunix) [reply src/backend/commands/tablecmds.c]: Thank you for your review. I have removed policy from MergeAttributes() signature and all mentions about updating policy in this function in a new commit. BTW, should I finally squash commits for this PR or it is ok to have a multiple ones?

File: src/backend/parser/parse_utilcmd.c (hunk 3 above)
Comment (darthunix) [reply src/backend/parser/parse_utilcmd.c]: It is possible to create these static helpers to add an item to a list if it is not present there yet. 
But it leads to duplication of functionality with `scr/backend/nodes/list.c` and I think it is not nice. For example
we already have [list_append_unique](https://github.com/greenplum-db/gpdb/blob/c960ebbe95542809b8a321e93c501a3d0fadc3ad/src/backend/nodes/list.c#L951) method that suits this task. But it requires changes in `list.c`, `nodes.h` and `equalfuncs.c` ... And in this case PR becomes not so non-invasive as I wish....
So should I create duplication with this static helpers or add `InheritColumn` structure to `NodeTag` type?

**Comment by darthunix** [comment]
BTW, should I finally squash commits for this PR or it is ok to have multiple ones?

File: src/backend/parser/parse_utilcmd.c (hunk 1 above)
Comment (darthunix) [reply src/backend/parser/parse_utilcmd.c]: Done

File: src/backend/parser/parse_utilcmd.c (hunk 2 above)
Comment (darthunix) [reply src/backend/parser/parse_utilcmd.c]: Done

File: src/backend/parser/parse_utilcmd.c [hunk 5]
```diff
@@ -83,7 +83,6 @@ typedef struct
 	List	   *grants;			/* GRANT items */
 } CreateSchemaStmtContext;
 
-
```
[the remark this answers is not in the document]
Comment (darthunix) [reply src/backend/parser/parse_utilcmd.c]: Done

**Comment by gkokolatos** [comment]
> BTW, should I finally squash commits for this PR or it is ok to have multiple ones?

Squashing is highly recommended.

File: src/backend/parser/parse_utilcmd.c (hunk 3 above)
Comment (gkokolatos) [reply src/backend/parser/parse_utilcmd.c]: Yes, you are absolutely correct. Adding such a function to list.c is really not nice. Exposing a new node `InheritColumn` to the whole codebase for such a task is also not nice. We are in agreement for those two parts and I appreciate your view point regarding invasiveness. 

How about adding in this file a static bool function to that returns the presence of a column on the list based on name? The signature can be `(List *, const str *)` if you so wish even and the `InheritColumn` can be specific knowledge of this function, i.e. in its body.
The same can be said for a static function to add a column to a list. And yes, combining those two would be add_unique yet in the patch at its presented form, finding and adding is not used an equal amount of times hence my suggestion.

Basically help readability by moving elsewhere the repetative moving parts.

Comment (darthunix) [reply src/backend/parser/parse_utilcmd.c]: Thanks for your advice. I've decided not to make a bool function returning a presence of a column in a list but rather to make a function returning a column of `InheritColumn` type or NULL from a list. It also allows to construct analogue of add_unique function but gives profit that we can through away `colindex` and store column index in `InheritColumn` list item itself.

Comment (gkokolatos) [reply src/backend/parser/parse_utilcmd.c]: Fantastic, that works too! I have two minor nitpicks if you allow me, I comment on the commit itself.

File: src/backend/parser/parse_utilcmd.c [hunk 6]
```diff
@@ -1729,33 +1765,14 @@ transformDistributedBy(ParseState *pstate, CreateStmtContext *cxt,
 					{
 						Form_pg_attribute inhattr = rel->rd_att->attrs[count];
 						char	   *inhname = NameStr(inhattr->attname);
-						bool		found = false;
+						InheritColumn *col = (InheritColumn *) palloc(sizeof(InheritColumn));
```
Comment (gkokolatos) [inline src/backend/parser/parse_utilcmd.c]: This call to `palloc()` may be superficial since the check of `attisdropped` might decide to not add it to the list.

Also to take it one step further if you allow me, `mergeColumnByName()` may not add the element to the list neither, if that element already exists. Why not move the responsibility for allocating the element in the `mergeColumnByName()` function?

**Comment by gkokolatos** [comment]
Thanks for your contribution and persistence!

I will push with some minor fixes in a branch under my permissions in order to go through the CI pipeline and then merge to 5X once ready.

Will link the branch here!

**Comment by gkokolatos** [comment]
Closed in favor of #8379
